tests/testInitialSetupConsent.php

```diff
@@ -0,0 +1,291 @@
+<?php
+/**
+ * Contains Class TestInitialSetupConsent.
+ *
+ * @package WP-Auth0
+ *
+ * @since 3.8.1
+ */
+
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Class TestInitialSetupConsent.
+ */
+class TestInitialSetupConsent extends TestCase {
+
+	use httpHelpers {
+		httpMock as protected httpMockDefault;
+	}
+
+	use RedirectHelpers;
+
+	use SetUpTestDb;
+
+	/**
+	 * Instance of WP_Auth0_Options.
+	 *
+	 * @var WP_Auth0_Options
+	 */
+	public static $opts;
+
+	/**
+	 * WP_Auth0_ErrorLog instance.
+	 *
+	 * @var WP_Auth0_ErrorLog
+	 */
+	protected static $error_log;
+
+	/**
+	 * Setup for entire test class.
+	 */
+	public static function setUpBeforeClass() {
+		parent::setUpBeforeClass();
+		self::$opts      = WP_Auth0_Options::Instance();
+		self::$error_log = new WP_Auth0_ErrorLog();
+	}
+
+	/**
+	 * Test that an invalid state is redirected to the right place.
+	 */
+	public function testThatInvalidStateRedirectsProperly() {
+		$this->startRedirectHalting();
+
+		$setup_consent = new WP_Auth0_InitialSetup_Consent( self::$opts );
+		$test_domain   = 'test-wp.auth0.com';
+		$test_token    = implode( '.', [ uniqid(), uniqid(), uniqid() ] );
+
+		$caught_redirect = [];
+		try {
+			$setup_consent->callback_with_token( $test_domain, $test_token, 'invalid_state' );
+		} catch ( Exception $e ) {
+			$caught_redirect = unserialize( $e->getMessage() );
+		}
+
+		$this->assertNotEmpty( $caught_redirect );
+		$this->assertEquals( 302, $caught_redirect['status'] );
+
+		$redirect_url = parse_url( $caught_redirect['location'] );
+
+		$this->assertEquals( '/wp-admin/admin.php', $redirect_url['path'] );
+		$this->assertContains( 'page=wpa0-setup', $redirect_url['query'] );
+		$this->assertContains( 'error=invalid_state', $redirect_url['query'] );
+
+		$this->assertEquals( $test_domain, self::$opts->get( 'domain' ) );
+		$this->assertEquals( $test_token, self::$opts->get( 'auth0_app_token' ) );
+
+		$this->assertEmpty( self::$error_log->get() );
+	}
+
+	/**
+	 * Test that the create client call is made.
+	 */
+	public function testThatClientCreationIsAttempted() {
+		$this->startHttpHalting();
+
+		$setup_consent = new WP_Auth0_InitialSetup_Consent( self::$opts );
+		$test_token    = implode( '.', [ uniqid(), uniqid(), uniqid() ] );
+
+		$caught_http = [];
+		try {
+			$setup_consent->callback_with_token( 'test-wp.auth0.com', $test_token, 'social' );
```

Comment:
how is this test logic different from the "test invalid state redirects"? Sounds like whatever caused the above test to fail should do the same here. What's changing? and if nothing is, why would an invalid_state scenario still make the create client request?

Reply:
`"social"` is one of two valid "state" values here so this proceeds to the API call